Tray has weird click behaviour

**Describe the bug**
Double clicking on tray icons does not close the tray popup again.
Triple clicking does but only if enough time passed.
Quadrupple clicking seems to always close it again x), or I'm too slow.

The other modules don't show this behavior and their popups go away on the second click as expected. (e.g. sound, time/calendar, notifications)

**To Reproduce**
Steps to reproduce the behavior:
1. Have tray module
2. Have something in the tray
3. Click fast twice on tray icon
4. Menu didn't close again

**Expected behavior**
Menu closes on second click.

**System information:**
 - Distro: arch
 - Compositor: sway 1.9
 - Ironbar version: ironbar 0.16.0-pre

**Configuration**

<details><summary>Config</summary>

```
anchor_to_edges = true
height = 42
position = "top"
icon_theme = "Papirus-Dark"

[[start]]
type = "workspaces"
all_monitors = false
on_scroll_up = "swaymsg workspace prev_on_output"
on_scroll_down = "swaymsg workspace next_on_output"

[[end]]
type = "music"
player_type = "mpd"
music_dir = "/home/merlijn/Music"

#[end.truncate]
#mode = "end"
#max_length = 100

[[end]]
type = "music"
player_type = "mpd"
truncate = "end"


[[end]]
type = "sys_info"
format = [
    "  {cpu_percent}% | {temp_c:k10temp-Tccd1}°C",
    "  {memory_used} / {memory_total} GB ({memory_percent}%)",
    "󰋊 {disk_used:/} / {disk_total:/} GB ({disk_percent:/}%)",
    "󰓢 {net_down:enp5s0} / {net_up:enp5s0} Mbps",
]

[end.interval]
memory = 30
cpu = 1
temps = 5
disks = 300
networks = 3

[[end]]
type = "volume"
format = "{icon} {percentage}%"
max_volume = 100

[end.icons]
volume_high = "󰕾"
volume_medium = "󰖀"
volume_low = "󰕿"
muted = "󰝟"

[end.truncate]
mode = "end"
length = 50


[[end]]
type = "clock"

[[end]]
type = "tray"
direction = "left_to_right"

[[end]]
type = "notifications"
show_count = true

[end.icons]
closed_none = "󰍥"
closed_some = "󱥂"
closed_dnd = "󱅯"
open_none = "󰍡"
open_some = "󱥁"
open_dnd = "󱅮"
```
</details>


<details><summary>Styles</summary>

```css
@define-color color_bg alpha(#2d2d2d, 0.7);
@define-color color_bg_dark alpha(#1c1c1c, 0.7);
@define-color color_border alpha(#424242, 0.7);
@define-color color_border_active alpha(#6699cc, 0.7);
@define-color color_text #ffffff;
@define-color color_urgent #8f0a0a;

/* -- base styles -- */

* {
    font-family: CaskaydiaCove Nerd Font, sans-serif;
    font-size: 16px;
    text-shadow: none;
    border: none;
    border-radius: 0;
    opacity: 1.0;
}

.background {
    background-color: rgba(0, 0, 0, 0);
}

button, label {
    box-shadow: none;
    background: none;
    background-color: rgba(0, 0, 0, 0);
    color: @color_text;
}

button:hover {
    background-color: @color_bg_dark;
}

scale trough {
    min-width: 1px;
    min-height: 2px;
}

#bar {
    border-top: 1px solid @color_border;
    background-color: @color_bg;
    background-image: none;
    box-shadow: none;
}

.popup {
    border: 1px solid @color_border;
    padding: 1em;
}


/* -- clipboard -- */

.clipboard {
    margin-left: 5px;
    font-size: 1.1em;
}

.popup-clipboard .item {
    padding-bottom: 0.3em;
    border-bottom: 1px solid @color_border;
}

.popup-clipboard {
    background-color: @color_bg;
}


/* -- clock -- */

.clock {
    font-weight: bold;
    margin-left: 5px;
}

.popup-clock .calendar-clock {
    color: @color_text;
    font-size: 2.5em;
    padding-bottom: 0.1em;
}

.popup-clock {
    background-color: @color_bg;
    color: @color_text;
}

.popup-clock .calendar .header {
    padding-top: 1em;
    border-top: 1px solid @color_border;
    font-size: 1.5em;
}

.popup-clock .calendar:selected {
    background-color: @color_border_active;
}

/* notifications */

.notifications .count {
    font-size: 0.6rem;
    background-color: @color_text;
    color: @color_bg;
    border-radius: 100%;
    margin-right: 3px;
    margin-top: 3px;
    padding-left: 4px;
    padding-right: 4px;
    opacity: 0.7;
}

/* -- script -- */

.script {
    padding-left: 10px;
}


/* -- sys_info -- */

.sysinfo {
    margin-left: 10px;
}

.sysinfo .item {
    margin-left: 5px;
}


/* -- tray -- */

.tray {
    background-color: rgba(0,0,0,0);
    margin-left: 10px;
}
.tray menu {
    background-color: rgba(0,0,0, 0.7);
    padding: 8px 0;
}
/* Overrides the label text just for the tray */
.tray label {
    /* color: black; */
}

/* -- volume -- */

.popup-volume .device-box {
    border-right: 1px solid @color_border;
}

.popup-volume {
    background-color: @color_bg;
}

/* -- power -- */

.popup-upower {
    background-color: @color_bg;
}

/* -- workspaces -- */

.workspaces .item.focused {
    box-shadow: inset 0 -3px;
    background-color: @color_bg_dark;
}

.workspaces .item:hover {
    box-shadow: inset 0 -3px;
}


```
</details>

> Finally getting round to clearing through the backlog. I'm not actually able to replicate this one, although I'm also not sure I can do anything about it from googling (I give GTK a sub-menu for the menu widget, it does the rest).
> 
> Are you able to demo the issue at all?

I'd happily demo it if I was able to install again
`cargo install ironbar`
```
error[E0432]: unresolved import `nix::fcntl::OFlag`
 --> /home/merlijn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/offer.rs:3:5
  |
3 | use nix::fcntl::OFlag;
  |     ^^^^^^^^^^^^^^^^^ no `OFlag` in `fcntl`

error[E0432]: unresolved import `nix::unistd::pipe2`
    --> /home/merlijn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/offer.rs:4:26
     |
4    | use nix::unistd::{close, pipe2};
     |                          ^^^^^
     |                          |
     |                          no `pipe2` in `unistd`
     |                          help: a similar name exists in the module: `pipe`
     |
note: found an item that was configured out
    --> /home/merlijn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/nix-0.27.1/src/unistd.rs:1233:8
     |
1233 | pub fn pipe2(flags: OFlag) -> Result<(RawFd, RawFd)> {
     |        ^^^^^
     = note: the item is gated behind the `fs` feature

error[E0432]: unresolved imports `nix::fcntl::fcntl`, `nix::fcntl::F_GETPIPE_SZ`, `nix::fcntl::F_SETPIPE_SZ`
   --> /home/merlijn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/wayland/wlr_data_control/mod.rs:13:18
    |
13  | use nix::fcntl::{fcntl, F_GETPIPE_SZ, F_SETPIPE_SZ};
    |                  ^^^^^  ^^^^^^^^^^^^  ^^^^^^^^^^^^ no `F_SETPIPE_SZ` in `fcntl`
    |                  |      |
    |                  |      no `F_GETPIPE_SZ` in `fcntl`
    |                  no `fcntl` in `fcntl`
    |
    = help: consider importing this module instead:
            nix::fcntl
note: found an item that was configured out
   --> /home/merlijn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/nix-0.27.1/src/fcntl.rs:509:8
    |
509 | pub fn fcntl(fd: RawFd, arg: FcntlArg) -> Result<c_int> {
    |        ^^^^^
    = note: the item is gated behind the `fs` feature

error[E0308]: mismatched types
   --> /home/merlijn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/compositor/hyprland.rs:156:51
    |
156 |                     let name = get_workspace_name(workspace_type);
    |                                ------------------ ^^^^^^^^^^^^^^ expected `WorkspaceType`, found `WorkspaceDestroyedEventData`
    |                                |
    |                                arguments to this function are incorrect
    |
note: function defined here
   --> /home/merlijn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ironbar-0.15.1/src/clients/compositor/hyprland.rs:247:4
    |
247 | fn get_workspace_name(name: WorkspaceType) -> String {
    |    ^^^^^^^^^^^^^^^^^^ -------------------

Some errors have detailed explanations: E0308, E0432.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `ironbar` (bin "ironbar") due to 4 previous errors
error: failed to compile `ironbar v0.15.1`, intermediate artifacts can be found at `/tmp/cargo-installQywNGM`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
cargo install ironbar  523.98s user 37.21s system 994% cpu 56.444 total
```

--- a/docs/modules/Clock.md
+++ b/docs/modules/Clock.md
@@ -8,12 +8,12 @@ Clicking on the widget opens a popup with the time and a calendar.
 
 > Type: `clock`
 
-| Name           | Type     | Default                            | Description                                                                         |
-|----------------|----------|------------------------------------|-------------------------------------------------------------------------------------|
-| `format`       | `string` | `%d/%m/%Y %H:%M`                   | Date/time format string. Pango markup is supported.                                 |
-| `format_popup` | `string` | `%H:%M:%S`                         | Date/time format string to display in the popup header. Pango markup is supported.  |
-| `locale`       | `string` | `$LC_TIME` or `$LANG` or `'POSIX'` | Locale to use (eg `en_GB`). Defaults to the system language (reading from env var). |
-| `orientation` | `'horizontal'` or `'vertical'` (shorthand: `'h'` or `'v'`) | `'horizontal'` | Orientation of the time on the clock button.                                                                                                      |
+| Name           | Type                                                       | Default                            | Description                                                                         |
+|----------------|------------------------------------------------------------|------------------------------------|-------------------------------------------------------------------------------------|
+| `format`       | `string`                                                   | `%d/%m/%Y %H:%M`                   | Date/time format string. Pango markup is supported.                                 |
+| `format_popup` | `string`                                                   | `%H:%M:%S`                         | Date/time format string to display in the popup header. Pango markup is supported.  |
+| `locale`       | `string`                                                   | `$LC_TIME` or `$LANG` or `'POSIX'` | Locale to use (eg `en_GB`). Defaults to the system language (reading from env var). |
+| `orientation`  | `'horizontal'` or `'vertical'` (shorthand: `'h'` or `'v'`) | `'horizontal'`                     | Orientation of the time on the clock button.                                        |
 
 > Detail on available tokens can be found here: <https://docs.rs/chrono/latest/chrono/format/strftime/index.html>
 

--- a/docs/modules/Tray.md
+++ b/docs/modules/Tray.md
@@ -6,11 +6,11 @@ Displays a fully interactive icon tray using the KDE `libappindicator` protocol.
 
 > Type: `tray`
 
-| Name                 | Type      | Default                                                         | Description                                                                                                                                                         |
-|----------------------|-----------|-----------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
-| `direction`          | `string`  | `left_to_right` if bar is horizontal, `top_to_bottom` otherwise | Direction to display the tray items. Possible values: `top_to_bottom`, `bottom_to_top`, `left_to_right`, `right_to_left`                                            |
-| `icon_size`          | `integer` | `16`                                                            | Size in pixels to display tray icons as.                                                                                                                            |
-| `prefer_theme_icons` | `bool`    | `true`                                                          | Requests that icons from the theme be used over the item-provided item. Most items only provide one or the other so this will have no effect in most circumstances. |
+| Name                 | Type                                                       | Default                 | Description                                                                                                                                                         |
+|----------------------|------------------------------------------------------------|-------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `orientation`        | `'horizontal'` or `'vertical'` (shorthand: `'h'` or `'v'`) | Matches bar orientation | The direction in which to pack tray icons.                                                                                                                          |
+| `icon_size`          | `integer`                                                  | `16`                    | Size in pixels to display tray icons as.                                                                                                                            |
+| `prefer_theme_icons` | `bool`                                                     | `true`                  | Requests that icons from the theme be used over the item-provided item. Most items only provide one or the other so this will have no effect in most circumstances. |
 
 <details>
 <summary>JSON</summary>

--- a/src/modules/tray/interface.rs
+++ b/src/modules/tray/interface.rs
@@ -1,11 +1,13 @@
+use glib::Propagation;
+use gtk::gdk::Gravity;
 use gtk::prelude::*;
-use gtk::{Image, Label, MenuItem};
+use gtk::{EventBox, Image, Label, MenuItem};
 use system_tray::item::{IconPixmap, StatusNotifierItem, Tooltip};
 
 /// Main tray icon to show on the bar
 pub(crate) struct TrayMenu {
+    pub event_box: EventBox,
     pub widget: MenuItem,
-    menu_widget: Option<system_tray::gtk_menu::Menu>,
     image_widget: Option<Image>,
     label_widget: Option<Label>,
 
@@ -17,12 +19,17 @@ pub(crate) struct TrayMenu {
 
 impl TrayMenu {
     pub fn new(item: StatusNotifierItem) -> Self {
+        let event_box = EventBox::new();
+
         let widget = MenuItem::new();
         widget.style_context().add_class("item");
+        event_box.add(&widget);
+
+        event_box.show_all();
 
         Self {
+            event_box,
             widget,
-            menu_widget: None,
             image_widget: None,
             label_widget: None,
             title: item.title,
@@ -99,7 +106,10 @@ impl TrayMenu {
     }
 
     pub fn set_menu_widget(&mut self, menu: system_tray::gtk_menu::Menu) {
-        self.widget.set_submenu(Some(&menu));
-        self.menu_widget = Some(menu);
+        self.event_box
+            .connect_button_press_event(move |event_box, _event| {
+                menu.popup_at_widget(event_box, Gravity::North, Gravity::South, None);
+                Propagation::Proceed
+            });
     }
 }

--- a/src/modules/tray/mod.rs
+++ b/src/modules/tray/mod.rs
@@ -2,12 +2,12 @@ mod icon;
 mod interface;
 
 use crate::clients::tray;
-use crate::config::CommonConfig;
+use crate::config::{CommonConfig, ModuleOrientation};
 use crate::modules::{Module, ModuleInfo, ModuleParts, ModuleUpdateEvent, WidgetContext};
 use crate::{glib_recv, lock, module_impl, send_async, spawn};
 use color_eyre::{Report, Result};
-use gtk::{prelude::*, PackDirection};
-use gtk::{IconTheme, MenuBar};
+use gtk::prelude::*;
+use gtk::{IconTheme, Orientation};
 use interface::TrayMenu;
 use serde::Deserialize;
 use std::collections::HashMap;
@@ -32,15 +32,13 @@ pub struct TrayModule {
     #[serde(default = "default_icon_size")]
     icon_size: u32,
 
-    /// Direction to display the tray items.
+    /// The direction in which to pack tray icons.
     ///
-    /// **Valid options**: `top_to_bottom`, `bottom_to_top`, `left_to_right`, `right_to_left`
+    /// **Valid options**: `horizontal`, `vertical`
     /// <br>
-    /// **Default**: `left_to_right` if bar is horizontal, `top_to_bottom` if bar is vertical
-    #[serde(default, deserialize_with = "deserialize_pack_direction")]
-    #[cfg_attr(feature = "schema", schemars(schema_with = "schema_pack_direction"))]
-    direction: Option<PackDirection>,
-
+    /// **Default**: `horizontal` for horizontal bars, `vertical` for vertical bars
+    #[serde(default)]
+    direction: Option<ModuleOrientation>,
     /// See [common options](module-level-options#common-options).
     #[serde(flatten)]
     pub common: Option<CommonConfig>,
@@ -50,36 +48,7 @@ const fn default_icon_size() -> u32 {
     16
 }
 
-fn deserialize_pack_direction<'de, D>(deserializer: D) -> Result<Option<PackDirection>, D::Error>
-where
-    D: serde::Deserializer<'de>,
-{
-    let value = Option::<String>::deserialize(deserializer)?;
-    value
-        .map(|v| match v.as_str() {
-            "left_to_right" => Ok(PackDirection::Ltr),
-            "right_to_left" => Ok(PackDirection::Rtl),
-            "top_to_bottom" => Ok(PackDirection::Ttb),
-            "bottom_to_top" => Ok(PackDirection::Btt),
-            _ => Err(serde::de::Error::custom("invalid value for direction")),
-        })
-        .transpose()
-}
-
-#[cfg(feature = "schema")]
-fn schema_pack_direction(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
-    use schemars::JsonSchema;
-    let mut schema: schemars::schema::SchemaObject = <String>::json_schema(gen).into();
-    schema.enum_values = Some(vec![
-        "top_to_bottom".into(),
-        "bottom_to_top".into(),
-        "left_to_right".into(),
-        "right_to_left".into(),
-    ]);
-    schema.into()
-}
-
-impl Module<MenuBar> for TrayModule {
+impl Module<gtk::Box> for TrayModule {
     type SendMessage = Event;
     type ReceiveMessage = ActivateRequest;
 
@@ -137,19 +106,17 @@ impl Module<MenuBar> for TrayModule {
         self,
         context: WidgetContext<Self::SendMessage, Self::ReceiveMessage>,
         info: &ModuleInfo,
-    ) -> Result<ModuleParts<MenuBar>> {
-        let container = MenuBar::new();
+    ) -> Result<ModuleParts<gtk::Box>> {
+        let orientation = self
+            .direction
+            .map(Orientation::from)
+            .unwrap_or(info.bar_position.orientation());
 
-        let direction = self.direction.unwrap_or(
-            if info.bar_position.orientation() == gtk::Orientation::Vertical {
-                PackDirection::Ttb
-            } else {
-                PackDirection::Ltr
-            },
-        );
-
-        container.set_pack_direction(direction);
-        container.set_child_pack_direction(direction);
+        // We use a `Box` here instead of the (supposedly correct) `MenuBar`
+        // as the latter has issues on Sway with menus focus-stealing from the bar.
+        //
+        // Each widget is wrapped in an EventBox, copying what Waybar does here.
+        let container = gtk::Box::new(orientation, 10);
 
         {
             let container = container.clone();
@@ -173,7 +140,7 @@ impl Module<MenuBar> for TrayModule {
 /// getting the diff since the previous update and applying it to the menu.
 fn on_update(
     update: Event,
-    container: &MenuBar,
+    container: &gtk::Box,
     menus: &mut HashMap<Box<str>, TrayMenu>,
     icon_theme: &IconTheme,
     icon_size: u32,
@@ -184,7 +151,7 @@ fn on_update(
             debug!("Received new tray item at '{address}': {item:?}");
 
             let mut menu_item = TrayMenu::new(*item);
-            container.add(&menu_item.widget);
+            container.pack_start(&menu_item.event_box, true, true, 0);
 
             if let Ok(image) = icon::get_image(&menu_item, icon_theme, icon_size, prefer_icons) {
                 menu_item.set_image(&image);

